crates/ruff_python_parser/src/parser/expression.rs

```diff
@@ -1898,7 +1912,11 @@ impl<'src> Parser<'src> {
         self.parse_comma_separated_list(RecoveryContextKind::ListElements, |parser| {
             elts.push(
                 parser
-                    .parse_star_expression_or_higher(AllowNamedExpression::Yes)
+                    .parse_named_expression_or_higher(
+                        ExpressionContext::default().with_starred_expression_allowed(
+                            StarredExpressionPrecedence::BitwiseOr,
+                        ),
+                    )
                     .expr,
```

Comment:
Nit: You could offer a helper to create the context: `ExpressionContext::starred_allowed(...)`.